src/libsync/propagatedownload.cpp

```diff
@@ -484,7 +486,22 @@ void PropagateDownloadFileQNAM::slotGetFinished()
         return;
     }
 
-    downloadFinished();
+    // do whatever is needed to add a checksum to the http upload request.
+    // in any case, the validator will emit signal startUpload to let the flow
+    // continue in slotStartUpload here.
+    _validator = new TransmissionChecksumValidator( _tmpFile.fileName() );
+    connect(_validator, SIGNAL(validated()), this, SLOT(downloadFinished()));
+    connect(_validator, SIGNAL(validationFailed(QString)), this, SLOT(slotChecksumFail(QString)));
+    _validator->downloadValidation(job->reply()->rawHeader(checkSumHeaderC));
+
+}
+
+void PropagateDownloadFileQNAM::slotChecksumFail( const QString& errMsg )
+{
+    _validator->deleteLater();
+    _tmpFile.remove();
```

Comment:
Yes, the _tmpFile is closed a couple of lines above, unconditionally. Should be good.